File: services/notebook-runtime-service/src/handlers/crud.rs
use axum::{
    Json,
    extract::{Path, Query, State},
    http::StatusCode,
    response::IntoResponse,
};
use uuid::Uuid;

use crate::AppState;
use crate::models::cell::{Cell, CreateCellRequest, UpdateCellRequest};
use crate::models::notebook::*;
use auth_middleware::layer::AuthUser;

// ── Notebook CRUD ──

pub async fn create_notebook(
    AuthUser(claims): AuthUser,
    State(state): State<AppState>,
    Json(body): Json<CreateNotebookRequest>,
) -> impl IntoResponse {
    let id = Uuid::now_v7();
    let description = body.description.unwrap_or_default();
    let kernel = body.default_kernel.unwrap_or_else(|| "python".to_string());

    let result = sqlx::query_as::<_, Notebook>(
        r#"INSERT INTO notebooks (id, name, description, owner_id, default_kernel)
           VALUES ($1, $2, $3, $4, $5)
           RETURNING *"#,
    )
    .bind(id)
    .bind(&body.name)
    .bind(&description)
    .bind(claims.sub)
    .bind(&kernel)
    .fetch_one(&state.db)
    .await;

    match result {
        Ok(nb) => (StatusCode::CREATED, Json(serde_json::json!(nb))).into_response(),
        Err(e) => {
            tracing::error!("create notebook: {e}");
            (StatusCode::INTERNAL_SERVER_ERROR, e.to_string()).into_response()
        }
    }
}

pub async fn list_notebooks(
    _user: AuthUser,
    State(state): State<AppState>,
    Query(params): Query<ListNotebooksQuery>,
) -> impl IntoResponse {
    let page = params.page.unwrap_or(1).max(1);
    let per_page = params.per_page.unwrap_or(20).clamp(1, 100);
    let offset = (page - 1) * per_page;
    let search = params.search.unwrap_or_default();
    let pattern = format!("%{search}%");

    let total: i64 = sqlx::query_scalar("SELECT COUNT(*) FROM notebooks WHERE name ILIKE $1")
        .bind(&pattern)
        .fetch_one(&state.db)
        .await
        .unwrap_or(0);

    let notebooks = sqlx::query_as::<_, Notebook>(
        r#"SELECT * FROM notebooks WHERE name ILIKE $1
           ORDER BY updated_at DESC LIMIT $2 OFFSET $3"#,
    )
    .bind(&pattern)
    .bind(per_page)
    .bind(offset)
    .fetch_all(&state.db)
    .await
    .unwrap_or_default();

    Json(
        serde_json::json!({ "data": notebooks, "total": total, "page": page, "per_page": per_page }),
    )
}

pub async fn get_notebook(
    _user: AuthUser,
    State(state): State<AppState>,
    Path(id): Path<Uuid>,
) -> impl IntoResponse {
    let notebook = sqlx::query_as::<_, Notebook>("SELECT * FROM notebooks WHERE id = $1")
        .bind(id)
        .fetch_optional(&state.db)
        .await;

    let cells = sqlx::query_as::<_, Cell>(
        "SELECT * FROM cells WHERE notebook_id = $1 ORDER BY position ASC",
    )
    .bind(id)
    .fetch_all(&state.db)
    .await
    .unwrap_or_default();

    match notebook {
        Ok(Some(nb)) => Json(serde_json::json!({ "notebook": nb, "cells": cells })).into_response(),
        Ok(None) => StatusCode::NOT_FOUND.into_response(),
        Err(e) => (StatusCode::INTERNAL_SERVER_ERROR, e.to_string()).into_response(),
    }
}

pub async fn update_notebook(
    _user: AuthUser,
    State(state): State<AppState>,
    Path(id): Path<Uuid>,
    Json(body): Json<UpdateNotebookRequest>,
) -> impl IntoResponse {
    let result = sqlx::query_as::<_, Notebook>(
        r#"UPDATE notebooks SET
           name = COALESCE($2, name),
           description = COALESCE($3, description),
           default_kernel = COALESCE($4, default_kernel),
           updated_at = NOW()
           WHERE id = $1 RETURNING *"#,
    )
    .bind(id)
    .bind(&body.name)
    .bind(&body.description)
    .bind(&body.default_kernel)
    .fetch_optional(&state.db)
    .await;

    match result {
        Ok(Some(nb)) => Json(serde_json::json!(nb)).into_response(),
        Ok(None) => StatusCode::NOT_FOUND.into_response(),
        Err(e) => (StatusCode::INTERNAL_SERVER_ERROR, e.to_string()).into_response(),
    }
}

pub async fn delete_notebook(
    _user: AuthUser,
    State(state): State<AppState>,
    Path(id): Path<Uuid>,
) -> impl IntoResponse {
    match sqlx::query("DELETE FROM notebooks WHERE id = $1")
        .bind(id)
        .execute(&state.db)
        .await
    {
        Ok(r) if r.rows_affected() > 0 => StatusCode::NO_CONTENT.into_response(),
        Ok(_) => StatusCode::NOT_FOUND.into_response(),
        Err(e) => (StatusCode::INTERNAL_SERVER_ERROR, e.to_string()).into_response(),
    }
}

// ── Cell CRUD ──

pub async fn add_cell(
    _user: AuthUser,
    State(state): State<AppState>,
    Path(notebook_id): Path<Uuid>,
    Json(body): Json<CreateCellRequest>,
) -> impl IntoResponse {
    let id = Uuid::now_v7();
    let cell_type = body.cell_type.unwrap_or_else(|| "code".to_string());
    let kernel = body.kernel.unwrap_or_else(|| "python".to_string());
    let source = body.source.unwrap_or_default();

    // Get next position
    let max_pos: Option<i32> =
        sqlx::query_scalar("SELECT MAX(position) FROM cells WHERE notebook_id = $1")
            .bind(notebook_id)
            .fetch_one(&state.db)
            .await
            .unwrap_or(None);

    let position = body.position.unwrap_or_else(|| max_pos.unwrap_or(0) + 1);

    let result = sqlx::query_as::<_, Cell>(
        r#"INSERT INTO cells (id, notebook_id, cell_type, kernel, source, position)
           VALUES ($1, $2, $3, $4, $5, $6)
           RETURNING *"#,
    )
    .bind(id)
    .bind(notebook_id)
    .bind(&cell_type)
    .bind(&kernel)
    .bind(&source)
    .bind(position)
    .fetch_one(&state.db)
    .await;

    match result {
        Ok(cell) => (StatusCode::CREATED, Json(serde_json::json!(cell))).into_response(),
        Err(e) => (StatusCode::INTERNAL_SERVER_ERROR, e.to_string()).into_response(),
    }
}

pub async fn update_cell(
    _user: AuthUser,
    State(state): State<AppState>,
    Path((_notebook_id, cell_id)): Path<(Uuid, Uuid)>,
    Json(body): Json<UpdateCellRequest>,
) -> impl IntoResponse {
    let result = sqlx::query_as::<_, Cell>(
        r#"UPDATE cells SET
           source = COALESCE($2, source),
           cell_type = COALESCE($3, cell_type),
           kernel = COALESCE($4, kernel),
           position = COALESCE($5, position),
           updated_at = NOW()
           WHERE id = $1 RETURNING *"#,
    )
    .bind(cell_id)
    .bind(&body.source)
    .bind(&body.cell_type)
    .bind(&body.kernel)
    .bind(body.position)
    .fetch_optional(&state.db)
    .await;

    match result {
        Ok(Some(c)) => Json(serde_json::json!(c)).into_response(),
        Ok(None) => StatusCode::NOT_FOUND.into_response(),
        Err(e) => (StatusCode::INTERNAL_SERVER_ERROR, e.to_string()).into_response(),
    }
}

pub async fn delete_cell(
    _user: AuthUser,
    State(state): State<AppState>,
    Path((_notebook_id, cell_id)): Path<(Uuid, Uuid)>,
) -> impl IntoResponse {
    match sqlx::query("DELETE FROM cells WHERE id = $1")
        .bind(cell_id)
        .execute(&state.db)
        .await
    {
        Ok(r) if r.rows_affected() > 0 => StatusCode::NO_CONTENT.into_response(),
        Ok(_) => StatusCode::NOT_FOUND.into_response(),
        Err(e) => (StatusCode::INTERNAL_SERVER_ERROR, e.to_string()).into_response(),
    }
}

File: openfoundry-go/services/notebook-runtime-service/internal/handler/crud.go
// Package handler — notebook + cell CRUD endpoints. 1:1 port of
// `services/notebook-runtime-service/src/handlers/crud.rs`.
//
// Wire format (paths / verbs / status codes / JSON shapes) matches
// the Rust router so the existing notebook frontends keep round-
// tripping. Behaviour highlights:
//
//   - `add_cell` reads `MAX(position)` to compute the next ordinal
//     when the caller doesn't pass one, mirroring the Rust path
//     exactly.
//   - `get_notebook` returns `{ "notebook": ..., "cells": [...] }`
//     so the client gets the cell list in one round-trip.
//   - Update endpoints implement the "COALESCE($n, column)" pattern
//     so PATCH semantics line up with the Rust handler (only fields
//     present in the body get rewritten).
package handler

import (
	"context"
	"encoding/json"
	"errors"
	"io"
	"net/http"
	"strconv"
	"strings"
	"time"

	"github.com/google/uuid"
	"github.com/jackc/pgx/v5"
	"github.com/jackc/pgx/v5/pgxpool"

	"github.com/openfoundry/openfoundry-go/services/notebook-runtime-service/internal/models"
)

// ── Notebook CRUD ───────────────────────────────────────────────────

// CreateNotebook mirrors `pub async fn create_notebook`.
func (s *State) CreateNotebook(w http.ResponseWriter, r *http.Request) {
	claims := requireClaims(w, r)
	if claims == nil {
		return
	}
	var body models.CreateNotebookRequest
	if err := decodeJSON(r, &body); err != nil {
		writeJSON(w, http.StatusBadRequest, errBody("invalid body"))
		return
	}
	description := ""
	if body.Description != nil {
		description = *body.Description
	}
	kernel := "python"
	if body.DefaultKernel != nil && *body.DefaultKernel != "" {
		kernel = *body.DefaultKernel
	}
	id, _ := uuid.NewV7()
	if s.Pool == nil {
		// Smoke-cluster fallback: synthesise the row the caller
		// would have got after a successful insert.
		now := time.Now().UTC()
		writeJSON(w, http.StatusCreated, models.Notebook{
			ID: id, Name: body.Name, Description: description,
			OwnerID: claims.Sub, DefaultKernel: kernel,
			CreatedAt: now, UpdatedAt: now,
		})
		return
	}
	row := s.Pool.QueryRow(r.Context(), `
        INSERT INTO notebooks (id, name, description, owner_id, default_kernel)
        VALUES ($1, $2, $3, $4, $5)
        RETURNING id, name, description, owner_id, default_kernel, created_at, updated_at`,
		id, body.Name, description, claims.Sub, kernel)
	nb, err := scanNotebook(row)
	if err != nil {
		writeJSON(w, http.StatusInternalServerError, errBody(err.Error()))
		return
	}
	writeJSON(w, http.StatusCreated, nb)
}

// ListNotebooks mirrors `pub async fn list_notebooks`.
func (s *State) ListNotebooks(w http.ResponseWriter, r *http.Request) {
	q := r.URL.Query()
	page := parseInt64(q.Get("page"), 1)
	if page < 1 {
		page = 1
	}
	perPage := parseInt64(q.Get("per_page"), 20)
	switch {
	case perPage < 1:
		perPage = 1
	case perPage > 100:
		perPage = 100
	}
	offset := (page - 1) * perPage
	pattern := "%" + q.Get("search") + "%"

	if s.Pool == nil {
		writeJSON(w, http.StatusOK, map[string]any{
			"data": []any{}, "total": 0, "page": page, "per_page": perPage,
		})
		return
	}
	var total int64
	_ = s.Pool.QueryRow(r.Context(),
		`SELECT COUNT(*) FROM notebooks WHERE name ILIKE $1`, pattern).Scan(&total)

	rows, err := s.Pool.Query(r.Context(), `
        SELECT id, name, description, owner_id, default_kernel, created_at, updated_at
        FROM notebooks WHERE name ILIKE $1
        ORDER BY updated_at DESC LIMIT $2 OFFSET $3`,
		pattern, perPage, offset)
	if err != nil {
		writeJSON(w, http.StatusInternalServerError, errBody(err.Error()))
		return
	}
	defer rows.Close()
	notebooks := []models.Notebook{}
	for rows.Next() {
		nb, err := scanNotebook(rows)
		if err != nil {
			writeJSON(w, http.StatusInternalServerError, errBody(err.Error()))
			return
		}
		notebooks = append(notebooks, nb)
	}
	writeJSON(w, http.StatusOK, map[string]any{
		"data": notebooks, "total": total, "page": page, "per_page": perPage,
	})
}

// GetNotebook mirrors `pub async fn get_notebook`. Returns the
// notebook plus its cells in one round-trip.
func (s *State) GetNotebook(w http.ResponseWriter, r *http.Request) {
	id, err := pathUUID(r, "notebook_id")
	if err != nil {
		writeJSON(w, http.StatusBadRequest, errBody("invalid notebook id"))
		return
	}
	if s.Pool == nil {
		writeJSON(w, http.StatusNotFound, nil)
		return
	}
	row := s.Pool.QueryRow(r.Context(), `
        SELECT id, name, description, owner_id, default_kernel, created_at, updated_at
        FROM notebooks WHERE id = $1`, id)
	nb, err := scanNotebook(row)
	if errors.Is(err, pgx.ErrNoRows) {
		writeJSON(w, http.StatusNotFound, nil)
		return
	}
	if err != nil {
		writeJSON(w, http.StatusInternalServerError, errBody(err.Error()))
		return
	}
	cells, err := loadCells(r.Context(), s.Pool, id)
	if err != nil {
		writeJSON(w, http.StatusInternalServerError, errBody(err.Error()))
		return
	}
	writeJSON(w, http.StatusOK, map[string]any{
		"notebook": nb,
		"cells":    cells,
	})
}

// UpdateNotebook mirrors `pub async fn update_notebook`. PATCH +
// PUT route through the same handler — the Rust crate registers them
// against the same axum function.
func (s *State) UpdateNotebook(w http.ResponseWriter, r *http.Request) {
	id, err := pathUUID(r, "notebook_id")
	if err != nil {
		writeJSON(w, http.StatusBadRequest, errBody("invalid notebook id"))
		return
	}
	var body models.UpdateNotebookRequest
	if err := decodeJSON(r, &body); err != nil {
		writeJSON(w, http.StatusBadRequest, errBody("invalid body"))
		return
	}
	if s.Pool == nil {
		writeJSON(w, http.StatusNotFound, nil)
		return
	}
	row := s.Pool.QueryRow(r.Context(), `
        UPDATE notebooks SET
            name = COALESCE($2, name),
            description = COALESCE($3, description),
            default_kernel = COALESCE($4, default_kernel),
            updated_at = NOW()
        WHERE id = $1
        RETURNING id, name, description, owner_id, default_kernel, created_at, updated_at`,
		id, body.Name, body.Description, body.DefaultKernel)
	nb, err := scanNotebook(row)
	if errors.Is(err, pgx.ErrNoRows) {
		writeJSON(w, http.StatusNotFound, nil)
		return
	}
	if err != nil {
		writeJSON(w, http.StatusInternalServerError, errBody(err.Error()))
		return
	}
	writeJSON(w, http.StatusOK, nb)
}

// DeleteNotebook mirrors `pub async fn delete_notebook`.
func (s *State) DeleteNotebook(w http.ResponseWriter, r *http.Request) {
	id, err := pathUUID(r, "notebook_id")
	if err != nil {
		writeJSON(w, http.StatusBadRequest, errBody("invalid notebook id"))
		return
	}
	if s.Pool == nil {
		w.WriteHeader(http.StatusNotFound)
		return
	}
	tag, err := s.Pool.Exec(r.Context(), `DELETE FROM notebooks WHERE id = $1`, id)
	if err != nil {
		writeJSON(w, http.StatusInternalServerError, errBody(err.Error()))
		return
	}
	if tag.RowsAffected() == 0 {
		w.WriteHeader(http.StatusNotFound)
		return
	}
	w.WriteHeader(http.StatusNoContent)
}

// ── Cell CRUD ───────────────────────────────────────────────────────

// AddCell mirrors `pub async fn add_cell`. When `position` is unset,
// inserts after the last cell — mirrors the Rust impl that reads
// `MAX(position)` first.
func (s *State) AddCell(w http.ResponseWriter, r *http.Request) {
	notebookID, err := pathUUID(r, "notebook_id")
	if err != nil {
		writeJSON(w, http.StatusBadRequest, errBody("invalid notebook id"))
		return
	}
	var body models.CreateCellRequest
	if err := decodeJSON(r, &body); err != nil {
		writeJSON(w, http.StatusBadRequest, errBody("invalid body"))
		return
	}
	cellType := "code"
	if body.CellType != nil && *body.CellType != "" {
		cellType = *body.CellType
	}
	kernel := "python"
	if body.Kernel != nil && *body.Kernel != "" {
		kernel = *body.Kernel
	}
	source := ""
	if body.Source != nil {
		source = *body.Source
	}
	id, _ := uuid.NewV7()
	if s.Pool == nil {
		position := int32(1)
		if body.Position != nil {
			position = *body.Position
		}
		now := time.Now().UTC()
		writeJSON(w, http.StatusCreated, models.Cell{
			ID:         id,
			NotebookID: notebookID,
			CellType:   cellType,
			Kernel:     kernel,
			Source:     source,
			Position:   position,
			CreatedAt:  now,
			UpdatedAt:  now,
		})
		return
	}
	position := int32(0)
	if body.Position != nil {
		position = *body.Position
	} else {
		var maxPos *int32
		_ = s.Pool.QueryRow(r.Context(),
			`SELECT MAX(position) FROM cells WHERE notebook_id = $1`,
			notebookID).Scan(&maxPos)
		if maxPos == nil {
			position = 1
		} else {
			position = *maxPos + 1
		}
	}
	row := s.Pool.QueryRow(r.Context(), `
        INSERT INTO cells (id, notebook_id, cell_type, kernel, source, position)
        VALUES ($1, $2, $3, $4, $5, $6)
        RETURNING id, notebook_id, cell_type, kernel, source, position,
                  last_output, execution_count, created_at, updated_at`,
		id, notebookID, cellType, kernel, source, position)
	cell, err := scanCell(row)
	if err != nil {
		writeJSON(w, http.StatusInternalServerError, errBody(err.Error()))
		return
	}
	writeJSON(w, http.StatusCreated, cell)
}

// UpdateCell mirrors `pub async fn update_cell`.
func (s *State) UpdateCell(w http.ResponseWriter, r *http.Request) {
	cellID, err := pathUUID(r, "cell_id")
	if err != nil {
		writeJSON(w, http.StatusBadRequest, errBody("invalid cell id"))
		return
	}
	var body models.UpdateCellRequest
	if err := decodeJSON(r, &body); err != nil {
		writeJSON(w, http.StatusBadRequest, errBody("invalid body"))
		return
	}
	if s.Pool == nil {
		writeJSON(w, http.StatusNotFound, nil)
		return
	}
	row := s.Pool.QueryRow(r.Context(), `
        UPDATE cells SET
            source = COALESCE($2, source),
            cell_type = COALESCE($3, cell_type),
            kernel = COALESCE($4, kernel),
            position = COALESCE($5, position),
            updated_at = NOW()
        WHERE id = $1
        RETURNING id, notebook_id, cell_type, kernel, source, position,
                  last_output, execution_count, created_at, updated_at`,
		cellID, body.Source, body.CellType, body.Kernel, body.Position)
	cell, err := scanCell(row)
	if errors.Is(err, pgx.ErrNoRows) {
		writeJSON(w, http.StatusNotFound, nil)
		return
	}
	if err != nil {
		writeJSON(w, http.StatusInternalServerError, errBody(err.Error()))
		return
	}
	writeJSON(w, http.StatusOK, cell)
}

// DeleteCell mirrors `pub async fn delete_cell`.
func (s *State) DeleteCell(w http.ResponseWriter, r *http.Request) {
	cellID, err := pathUUID(r, "cell_id")
	if err != nil {
		writeJSON(w, http.StatusBadRequest, errBody("invalid cell id"))
		return
	}
	if s.Pool == nil {
		w.WriteHeader(http.StatusNotFound)
		return
	}
	tag, err := s.Pool.Exec(r.Context(), `DELETE FROM cells WHERE id = $1`, cellID)
	if err != nil {
		writeJSON(w, http.StatusInternalServerError, errBody(err.Error()))
		return
	}
	if tag.RowsAffected() == 0 {
		w.WriteHeader(http.StatusNotFound)
		return
	}
	w.WriteHeader(http.StatusNoContent)
}

// ── Repository helpers ──────────────────────────────────────────────

// rowScanner abstracts pgx.Row + pgx.Rows so scan helpers work for
// either a one-row QueryRow result or a streaming Query result.
type rowScanner interface {
	Scan(dest ...any) error
}

func scanNotebook(s rowScanner) (models.Notebook, error) {
	var nb models.Notebook
	err := s.Scan(&nb.ID, &nb.Name, &nb.Description, &nb.OwnerID,
		&nb.DefaultKernel, &nb.CreatedAt, &nb.UpdatedAt)
	return nb, err
}

func scanCell(s rowScanner) (models.Cell, error) {
	var c models.Cell
	var lastOutput []byte
	err := s.Scan(&c.ID, &c.NotebookID, &c.CellType, &c.Kernel, &c.Source,
		&c.Position, &lastOutput, &c.ExecutionCount, &c.CreatedAt, &c.UpdatedAt)
	if len(lastOutput) > 0 {
		c.LastOutput = lastOutput
	}
	return c, err
}

func loadCells(ctx context.Context, pool *pgxpool.Pool, notebookID uuid.UUID) ([]models.Cell, error) {
	rows, err := pool.Query(ctx, `
        SELECT id, notebook_id, cell_type, kernel, source, position,
               last_output, execution_count, created_at, updated_at
        FROM cells WHERE notebook_id = $1 ORDER BY position ASC`, notebookID)
	if err != nil {
		return nil, err
	}
	defer rows.Close()
	out := []models.Cell{}
	for rows.Next() {
		c, err := scanCell(rows)
		if err != nil {
			return nil, err
		}
		out = append(out, c)
	}
	return out, nil
}

// ── Misc helpers shared with sessions.go ────────────────────────────

func parseInt64(s string, fallback int64) int64 {
	if s == "" {
		return fallback
	}
	n, err := strconv.ParseInt(strings.TrimSpace(s), 10, 64)
	if err != nil {
		return fallback
	}
	return n
}

func decodeJSON(r *http.Request, dst any) error {
	if r.Body == nil || r.ContentLength == 0 {
		return nil
	}
	defer r.Body.Close()
	if err := json.NewDecoder(r.Body).Decode(dst); err != nil && !errors.Is(err, io.EOF) {
		return err
	}
	return nil
}
